lib/shred.ts

```diff
@@ -124,19 +136,19 @@ function shredRecordInternal(fields, record, data, rlvl, dlvl) {
     for (let i = 0; i < values.length; ++i) {
       const rlvl_i = i === 0 ? rlvl : field.rLevelMax;
 
-      if (field.isNested) {
+      if (field.isNested && isDefined(field.fields)) {
         shredRecordInternal(
             field.fields,
-            values[i],
+            values[i] as Record<string, unknown>,
             data,
             rlvl_i,
             field.dLevelMax);
       } else {
-        data[field.path].distinct_values.add(values[i]);
-        data[field.path].values.push(parquet_types.toPrimitive(fieldType, values[i]));
-        data[field.path].rlevels.push(rlvl_i);
-        data[field.path].dlevels.push(field.dLevelMax);
-        data[field.path].count += 1;
+        data[path].distinct_values.add(values[i]);
+        data[path].values.push(parquet_types.toPrimitive(fieldType as string, values[i]));
```

Comment:
I don't like covering up the undefined case here which would then pass off the error to `toPrimitive`. (same down below with `fromPrimitive`)

Option 1: Allow `toPrimitive` and `fromPrimitive` to accept undefined and handle the error there.

```typescript
export function toPrimitive(type: string | undefined, value: unknown) {
  if (type === undefined || !(type in PARQUET_LOGICAL_TYPES)) {
    throw 'invalid type: ' + type || "undefined";
  }
// ...
```

Option 2: Just do an `|| "undefined"` here:
```suggestion
        data[path].values.push(parquet_types.toPrimitive(fieldType || "undefined", values[i]));
```

I lean a bit more toward Option 1, because an "||" here could do strange things if the setup changes over in types.ts in the future.

Thoughts?

Reply:
I followed your advice and went with option 1 to have toPrimitve deal with undefined types